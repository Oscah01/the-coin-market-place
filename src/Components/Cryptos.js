import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import { BsArrowRightCircle } from 'react-icons/bs';
import { fetchCryptos } from '../redux/cryptocurrencies/cryptos';
import Search from './Search';
import Footer from './Footer';

const Cryptos = () => {
  const dispatch = useDispatch();
  const cryptos = useSelector((state) => state);
  useEffect(() => {
    if (!cryptos.length) {
      dispatch(fetchCryptos());
    }
  }, [cryptos.length, dispatch]);

  const handleId = (e) => {
    localStorage.clear();
    const detail = cryptos.find((crypto) => crypto.symbol === e.target.id);
    localStorage.setItem('detail', JSON.stringify(detail));
  };

  const [param, setSearch] = useState('');
  const handleSearch = (e) => {
    const query = e.target.value.trim();
    setSearch(query);
  };

  const filteredCryptos = cryptos.filter((crypto) => crypto.name.replace(/-/, '').toLowerCase().includes(param));

  return (
    <>
      <h1 style={{ color: '#fff', textAlign: 'center' }}>Crypto market place</h1>
      <Search param={param} handleSearch={handleSearch} />
      {!filteredCryptos.length && <div className="loader" />}
      {filteredCryptos.length && (
        <div
          style={{
            backgroundImage: `url(${cryptos[31].image.large})`,
          }}
          className="all-coins"
        >
          <NavLink
            to={`/Details/${cryptos[31].id}`}
            style={{
              position: 'absolute',
              top: '2px',
              right: '9px',
              color: 'orange',
              fontSize: '30px',
            }}
          >
            <BsArrowRightCircle id={cryptos[31].symbol} onClick={handleId} />
          </NavLink>
          <div className="currency-details">
            <p className="currency-price">
              <span> Current Price:</span>
              <span>
                $
                {cryptos[31].market_data.current_price.usd}
              </span>
            </p>
            <span>{cryptos[31].name}</span>
          </div>
        </div>
      )}
      <div className="coins-container">
        {filteredCryptos.map((currency) => (
          <div key={currency.symbol} className="coins">
            <div
              style={{
                backgroundImage: `url(${currency.image.large})`,
              }}
            >
              <p style={{ position: 'absolute', top: '14px', left: '9px' }}>
                {currency.symbol.toUpperCase()}
              </p>
              <div className="currency-details">
                <p className="currency-price">
                  <span> Current Price:</span>
                  <span>
                    $
                    {currency.market_data.current_price.usd}
                  </span>
                </p>
                <span>{currency.name}</span>
              </div>
              <NavLink
                to={`/Details/${currency.id}`}
                style={{
                  position: 'absolute',
                  top: '2px',
                  right: '9px',
                  color: 'orange',
                  fontSize: '25px',
                }}
              >
                <BsArrowRightCircle id={currency.symbol} onClick={handleId} />
              </NavLink>
            </div>
          </div>
        ))}
      </div>
      <Footer />
    </>
  );
};

export default Cryptos;
