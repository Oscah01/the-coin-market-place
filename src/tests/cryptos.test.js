import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../redux/configureStore';
import Cryptos from '../Components/Cryptos';
import currenciesReducer, { fetchCryptos } from '../redux/cryptocurrencies/cryptos';

describe('Cryptos', () => {
  it('should render the Cryptos component', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Cryptos />
        </BrowserRouter>
      </Provider>,
    );
    expect(Cryptos).toMatchSnapshot();
  });
});

describe('The reducer', () => {
  test('Should update the state with API data', () => {
    const coins = jest.mock();
    expect(currenciesReducer(coins, fetchCryptos(coins))).toEqual(coins);
  });
});
