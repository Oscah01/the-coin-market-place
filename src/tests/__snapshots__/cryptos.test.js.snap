// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cryptos should render the Cryptos component 1`] = `[Function]`;
